compiler/language_server/src/features_candy/analyzer/mod.rs

```diff
@@ -73,33 +76,42 @@ pub async fn run_server(
                 Err(TryRecvError::Disconnected) => break 'server_loop,
             };
             match event {
-                Event::UpdateModule(module, content) => {
+                Message::UpdateModule(module, content) => {
                     db.did_change_module(&module, content);
                     outgoing_hints.send(module.clone(), vec![]).await;
-                    hints_finders
+                    analyzers
                         .entry(module.clone())
                         .and_modify(|it| it.module_changed())
-                        .or_insert_with(|| HintsFinder::for_module(module.clone()));
+                        .or_insert_with(|| ModuleAnalyzer::for_module(module.clone()));
                 }
-                Event::CloseModule(module) => {
+                Message::CloseModule(module) => {
                     db.did_close_module(&module);
-                    hints_finders.remove(&module);
+                    analyzers.remove(&module);
                 }
-                Event::Shutdown => {
+                Message::Shutdown => {
                     incoming_events.close();
                 }
             }
         }
 
-        let Some(module) = hints_finders.keys().choose(&mut thread_rng()).cloned() else {
+        let Some(module) = analyzers.keys().choose(&mut thread_rng()).cloned() else {
             client.update_status(None);
             continue;
         };
-        let hints_finder = hints_finders.get_mut(&module).unwrap();
+        let analyzer = analyzers.get_mut(&module).unwrap();
+
+        analyzer.run(&db, &client).await;
 
-        hints_finder.run(&db, &client).await;
+        let insights = analyzer.insights(&db);
+        let mut diagnostics = vec![];
+        let mut hints = vec![];
+        for insight in insights {
+            match insight {
+                Insight::Diagnostic(diagnostic) => diagnostics.push(diagnostic),
+                Insight::Hint(hint) => hints.push(hint),
+            }
+        }
```

Comment:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.partition_map

```suggestion
        let (diagnostics, mut hints): (Vec<_>, Vec<_>) =
            insights.into_iter().partition_map(|it| match it {
                Insight::Diagnostic(diagnostic) => Either::Left(diagnostic),
                Insight::Hint(hint) => Either::Right(hint),
            });
```